node_settled_handler.go

```diff
@@ -115,25 +110,60 @@ func (p *NodeSettledHandler) subscribeEvents(ctx context.Context) error {
 		if !ok {
 			continue
 		}
-		if !finalEvent.FinalHtlcEvent.Settled {
-			// TODO: Handle final fail?
-
-			continue
-		}
 
 		key := types.CircuitKey{
 			ChanID: event.IncomingChannelId,
 			HtlcID: event.IncomingHtlcId,
 		}
 
-		if err := p.handleFinalHtlc(ctx, key); err != nil {
+		failedResolution := !finalEvent.FinalHtlcEvent.Settled
+
+		if err := p.handleFinalHtlc(ctx, key, failedResolution); err != nil {
 			return err
 		}
 	}
 }
 
 func (p *NodeSettledHandler) handleFinalHtlc(ctx context.Context,
-	key types.CircuitKey) error {
+	key types.CircuitKey, failedResolution bool) error {
+
+	var (
+		err         error
+		invoiceHash *lntypes.Hash
+		status      types.Status
+	)
+	if failedResolution {
+		invoiceHash, err = p.handleFinalFailedHtlc(ctx, key)
+		if err != nil {
+			return err
+		}
+
+		status = types.StatusFailed
+	} else {
+		invoiceHash, err = p.handleFinalSettledHtlc(ctx, key)
+		if err != nil {
+			return err
+		}
+
+		status = types.StatusSettled
```

Comment:
If this is just for logging, maybe you can make a logger at the top of this function using `logger.With`, and then log inside the if blocks `Infow("final final received")` and `Infow("final settle received")`?